# Machine Learning Assignment 2

## Bank Marketing Classification using Multiple Machine Learning Models

### Student Details

- **Name:** KARUNKUMAR B V
- **BITS ID:** 2025AC05811
- **BITS Email ID:** 2025AC05811@wilp.bits-pilani.ac.in

---

## Objective

The objective of this assignment is to build and compare multiple Machine Learning classification models using the Bank Marketing dataset. The models will be evaluated using various performance metrics including Accuracy, AUC Score, Precision, Recall, F1 Score, and Matthews Correlation Coefficient (MCC).

The trained models will later be integrated into a Streamlit web application for interactive prediction and deployed on Streamlit Community Cloud.

In [ ]:
# =============================================================================
# Import Required Libraries
# =============================================================================

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

print("Libraries imported successfully.")

Libraries imported successfully.


In [ ]:
# =============================================================================
# Cell 2: Load the Dataset
# =============================================================================

df = pd.read_csv("bank-full.csv", sep=";")

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [ ]:
# =============================================================================
# Step 1: Explore the Dataset Structure
# =============================================================================

# Display the number of rows and columns in the dataset
print("Dataset Shape:", df.shape)

# Display the column names
print("\nColumn Names:")
print(df.columns)

# Display information about the dataset
print("\nDataset Information:")
df.info()

Dataset Shape: (45211, 17)

Column Names:
Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'y'],
      dtype='object')

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null

In [ ]:
# =============================================================================
# Cell 3: Display the First 5 Rows
# =============================================================================

df.head()

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no


# Step 2: Machine Learning Classification Models and Evaluation

## 2.1 Data Preprocessing

Before training the Machine Learning classification models, the dataset is preprocessed to ensure it is suitable for analysis. The preprocessing activities include understanding the dataset, checking for missing values, encoding categorical variables, splitting the dataset into training and testing sets, and applying feature scaling where required.

These preprocessing steps prepare the dataset for building and evaluating the required Machine Learning models.

In [ ]:
# =============================================================================
# 2.1.1 Statistical Summary of Numerical Features
# =============================================================================

# Display summary statistics for all numerical columns
df.describe()

,age,balance,day,duration,campaign,pdays,previous
count,45211.000000,45211.000000,45211.000000,45211.000000,45211.000000,45211.000000,45211.000000
mean,40.936210,1362.272058,15.806419,258.163080,2.763841,40.197828,0.580323
std,10.618762,3044.765829,8.322476,257.527812,3.098021,100.128746,2.303441
min,18.000000,-8019.000000,1.000000,0.000000,1.000000,-1.000000,0.000000
25%,33.000000,72.000000,8.000000,103.000000,1.000000,-1.000000,0.000000
50%,39.000000,448.000000,16.000000,180.000000,2.000000,-1.000000,0.000000
75%,48.000000,1428.000000,21.000000,319.000000,3.000000,-1.000000,0.000000
max,95.000000,102127.000000,31.000000,4918.000000,63.000000,871.000000,275.000000


In [ ]:
# =============================================================================
# 2.1.2 Check for Missing Values
# =============================================================================

# Display the number of missing values in each column
missing_values = df.isnull().sum()

print("Missing Values in Each Column:")
print(missing_values)

Missing Values in Each Column:
age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64


In [ ]:
# =============================================================================
# 2.1.3 Analyse the Target Variable
# =============================================================================

# Display the count of each class in the target variable
print("Target Variable Distribution:\n")
print(df['y'].value_counts())

Target Variable Distribution:

y
no     39922
yes     5289
Name: count, dtype: int64


In [ ]:
# =============================================================================
# 2.1.4 Encode Categorical Variables
# =============================================================================

from sklearn.preprocessing import LabelEncoder

# Create a LabelEncoder object
label_encoder = LabelEncoder()

# Identify all categorical columns
categorical_columns = df.select_dtypes(include='object').columns

# Apply Label Encoding to each categorical column
for column in categorical_columns:
    df[column] = label_encoder.fit_transform(df[column])

# Display the first five records after encoding
df.head()

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,4,1,2,0,2143,1,0,2,5,8,261,1,-1,0,3,0
1,44,9,2,1,0,29,1,0,2,5,8,151,1,-1,0,3,0
2,33,2,1,1,0,2,1,1,2,5,8,76,1,-1,0,3,0
3,47,1,1,3,0,1506,1,0,2,5,8,92,1,-1,0,3,0
4,33,11,2,3,0,1,0,0,2,5,8,198,1,-1,0,3,0


In [ ]:
# =============================================================================
# 2.1.5 Split the Dataset into Features and Target Variable
# =============================================================================

# Separate the input features (X) and target variable (y)
X = df.drop('y', axis=1)
y = df['y']
print("Training Feature Order:")
print(X.columns.tolist())

# Display the shape of the feature matrix and target vector
print("Feature Matrix (X) Shape :", X.shape)
print("Target Variable (y) Shape:", y.shape)

Training Feature Order:
['age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome']
Feature Matrix (X) Shape : (45211, 16)
Target Variable (y) Shape: (45211,)


In [ ]:
# =============================================================================
# 2.1.6 Split the Dataset into Training and Testing Sets
# =============================================================================

from sklearn.model_selection import train_test_split

# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

# Display the shape of the training and testing datasets
print("Training Feature Set :", X_train.shape)
print("Testing Feature Set  :", X_test.shape)
print("Training Target Set  :", y_train.shape)
print("Testing Target Set   :", y_test.shape)

Training Feature Set : (36168, 16)
Testing Feature Set  : (9043, 16)
Training Target Set  : (36168,)
Testing Target Set   : (9043,)


In [ ]:
# =============================================================================
# 2.1.7 Feature Scaling
# =============================================================================

from sklearn.preprocessing import StandardScaler

# Create a StandardScaler object
scaler = StandardScaler()

# Fit the scaler on the training data and transform it
X_train_scaled = scaler.fit_transform(X_train)

# Use the same scaler to transform the testing data
X_test_scaled = scaler.transform(X_test)

print("Feature scaling completed successfully.")

Feature scaling completed successfully.


## 2.2 Logistic Regression

Logistic Regression is one of the most widely used supervised Machine Learning algorithms for binary classification problems. It learns the relationship between the input features and the target variable and predicts the probability that a customer belongs to a particular class.

In this assignment, Logistic Regression is used to predict whether a customer will subscribe to a term deposit based on the customer attributes and previous marketing campaign information.

In [ ]:
# =============================================================================
# 2.2.1 Train the Logistic Regression Model
# =============================================================================

from sklearn.linear_model import LogisticRegression

# Create the Logistic Regression model
logistic_model = LogisticRegression(random_state=42)

# Train the model using the training dataset
logistic_model.fit(X_train_scaled, y_train)

print("Logistic Regression model trained successfully.")

Logistic Regression model trained successfully.


In [ ]:
# =============================================================================
# 2.2.2 Predict the Test Dataset
# =============================================================================

# Predict the target values for the testing dataset
y_pred = logistic_model.predict(X_test_scaled)

print("Prediction completed successfully.")

Prediction completed successfully.


In [ ]:
# =============================================================================
# 2.2.3 Evaluate Logistic Regression - Accuracy
# =============================================================================

from sklearn.metrics import accuracy_score

# Calculate the accuracy of the Logistic Regression model
logistic_accuracy = accuracy_score(y_test, y_pred)

print(f"Logistic Regression Accuracy: {logistic_accuracy:.4f}")

Logistic Regression Accuracy: 0.8879


In [ ]:
# =============================================================================
# 2.2.4 Evaluate Logistic Regression - Precision, Recall and F1 Score
# =============================================================================

from sklearn.metrics import precision_score, recall_score, f1_score

# Calculate evaluation metrics
logistic_precision = precision_score(y_test, y_pred)
logistic_recall = recall_score(y_test, y_pred)
logistic_f1 = f1_score(y_test, y_pred)

# Display the evaluation metrics
print(f"Precision : {logistic_precision:.4f}")
print(f"Recall    : {logistic_recall:.4f}")
print(f"F1 Score  : {logistic_f1:.4f}")

Precision : 0.5970
Recall    : 0.2172
F1 Score  : 0.3185


In [ ]:
# =============================================================================
# 2.2.5 Evaluate Logistic Regression - AUC Score and MCC
# =============================================================================

from sklearn.metrics import roc_auc_score, matthews_corrcoef

# Calculate prediction probabilities for the positive class
y_prob = logistic_model.predict_proba(X_test_scaled)[:, 1]

# Calculate AUC Score
logistic_auc = roc_auc_score(y_test, y_prob)

# Calculate Matthews Correlation Coefficient (MCC)
logistic_mcc = matthews_corrcoef(y_test, y_pred)

# Display the evaluation metrics
print(f"AUC Score : {logistic_auc:.4f}")
print(f"MCC Score : {logistic_mcc:.4f}")

AUC Score : 0.8700
MCC Score : 0.3134


The Logistic Regression model achieved a good overall accuracy (88.79%). However, the recall was relatively low (21.72%), indicating that the model identified only a small proportion of customers who actually subscribed to the term deposit. This suggests that although the model performed well overall, it was less effective in detecting the minority class.

## 2.3 Decision Tree Classifier

Decision Tree is a supervised Machine Learning algorithm used for both classification and regression problems. It makes predictions by recursively splitting the dataset into smaller subsets based on the most informative features, forming a tree-like structure of decision rules. Unlike Logistic Regression, Decision Tree does not require feature scaling because it makes decisions based on feature values rather than distances.

In this assignment, the Decision Tree classifier is trained to predict whether a customer will subscribe to a term deposit based on customer information and previous marketing campaign details.

In [ ]:
# =============================================================================
# 2.3.1 Train the Decision Tree Classifier
# =============================================================================

from sklearn.tree import DecisionTreeClassifier

# Create the Decision Tree model
decision_tree_model = DecisionTreeClassifier(random_state=42)

# Train the model using the training dataset
decision_tree_model.fit(X_train, y_train)

print("Decision Tree model trained successfully.")

Decision Tree model trained successfully.


In [ ]:
# =============================================================================
# 2.3.2 Predict and Evaluate the Decision Tree Classifier
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    matthews_corrcoef
)

# Predict the target values for the testing dataset
dt_y_pred = decision_tree_model.predict(X_test)

# Predict probabilities for AUC calculation
dt_y_prob = decision_tree_model.predict_proba(X_test)[:, 1]

# Calculate evaluation metrics
dt_accuracy = accuracy_score(y_test, dt_y_pred)
dt_precision = precision_score(y_test, dt_y_pred)
dt_recall = recall_score(y_test, dt_y_pred)
dt_f1 = f1_score(y_test, dt_y_pred)
dt_auc = roc_auc_score(y_test, dt_y_prob)
dt_mcc = matthews_corrcoef(y_test, dt_y_pred)

# Display the evaluation metrics
print("Decision Tree Performance Metrics")
print("-" * 40)
print(f"Accuracy : {dt_accuracy:.4f}")
print(f"Precision: {dt_precision:.4f}")
print(f"Recall   : {dt_recall:.4f}")
print(f"F1 Score : {dt_f1:.4f}")
print(f"AUC Score: {dt_auc:.4f}")
print(f"MCC Score: {dt_mcc:.4f}")

Decision Tree Performance Metrics
----------------------------------------
Accuracy : 0.8740
Precision: 0.4782
Recall   : 0.4821
F1 Score : 0.4801
AUC Score: 0.7050
MCC Score: 0.4085


## 2.4 K-Nearest Neighbour (KNN) Classifier

K-Nearest Neighbour (KNN) is a supervised Machine Learning algorithm used for classification and regression problems. Unlike Logistic Regression and Decision Tree, KNN does not build an explicit model during training. Instead, it classifies a new data point by identifying the 'K' nearest training samples based on distance and assigning the class that occurs most frequently among those neighbours.

In this assignment, the K-Nearest Neighbour classifier is used to predict whether a customer will subscribe to a term deposit based on the similarity of customer attributes. Since KNN is a distance-based algorithm, feature scaling is essential to ensure that all features contribute equally to the distance calculation.

In [ ]:
# =============================================================================
# 2.4.1 Train the K-Nearest Neighbour (KNN) Classifier
# =============================================================================

from sklearn.neighbors import KNeighborsClassifier

# Create the KNN classifier with K = 5
knn_model = KNeighborsClassifier(n_neighbors=5)

# Train the model using the scaled training dataset
knn_model.fit(X_train_scaled, y_train)

print("K-Nearest Neighbour model trained successfully.")

K-Nearest Neighbour model trained successfully.


In [ ]:
# =============================================================================
# 2.4.2 Predict and Evaluate the K-Nearest Neighbour (KNN) Classifier
# =============================================================================

# Predict the target values for the testing dataset
knn_y_pred = knn_model.predict(X_test_scaled)

# Predict probabilities for AUC calculation
knn_y_prob = knn_model.predict_proba(X_test_scaled)[:, 1]

# Calculate evaluation metrics
knn_accuracy = accuracy_score(y_test, knn_y_pred)
knn_precision = precision_score(y_test, knn_y_pred)
knn_recall = recall_score(y_test, knn_y_pred)
knn_f1 = f1_score(y_test, knn_y_pred)
knn_auc = roc_auc_score(y_test, knn_y_prob)
knn_mcc = matthews_corrcoef(y_test, knn_y_pred)

# Display the evaluation metrics
print("K-Nearest Neighbour Performance Metrics")
print("-" * 45)
print(f"Accuracy : {knn_accuracy:.4f}")
print(f"Precision: {knn_precision:.4f}")
print(f"Recall   : {knn_recall:.4f}")
print(f"F1 Score : {knn_f1:.4f}")
print(f"AUC Score: {knn_auc:.4f}")
print(f"MCC Score: {knn_mcc:.4f}")

K-Nearest Neighbour Performance Metrics
---------------------------------------------
Accuracy : 0.8912
Precision: 0.5859
Recall   : 0.3346
F1 Score : 0.4259
AUC Score: 0.8260
MCC Score: 0.3885


## 2.5 Naive Bayes Classifier

Naive Bayes is a supervised Machine Learning algorithm based on Bayes' Theorem. It is widely used for classification problems and predicts the class of a data instance by calculating the probability of each class given the input features. The algorithm assumes that all input features are independent of each other, which is known as the "naive" assumption.

In this assignment, the Gaussian Naive Bayes classifier is used to predict whether a customer will subscribe to a term deposit based on customer information and previous marketing campaign details. Although the independence assumption may not always hold in real-world datasets, Naive Bayes is computationally efficient and often provides good baseline classification performance.

In [ ]:
# =============================================================================
# 2.5.1 Train the Gaussian Naive Bayes Classifier
# =============================================================================

from sklearn.naive_bayes import GaussianNB

# Create the Gaussian Naive Bayes model
naive_bayes_model = GaussianNB()

# Train the model using the training dataset
naive_bayes_model.fit(X_train, y_train)

print("Gaussian Naive Bayes model trained successfully.")

Gaussian Naive Bayes model trained successfully.


In [ ]:
# =============================================================================
# 2.5.2 Predict and Evaluate the Gaussian Naive Bayes Classifier
# =============================================================================

# Predict the target values for the testing dataset
nb_y_pred = naive_bayes_model.predict(X_test)

# Predict probabilities for AUC calculation
nb_y_prob = naive_bayes_model.predict_proba(X_test)[:, 1]

# Calculate evaluation metrics
nb_accuracy = accuracy_score(y_test, nb_y_pred)
nb_precision = precision_score(y_test, nb_y_pred)
nb_recall = recall_score(y_test, nb_y_pred)
nb_f1 = f1_score(y_test, nb_y_pred)
nb_auc = roc_auc_score(y_test, nb_y_prob)
nb_mcc = matthews_corrcoef(y_test, nb_y_pred)

# Display the evaluation metrics
print("Gaussian Naive Bayes Performance Metrics")
print("-" * 45)
print(f"Accuracy : {nb_accuracy:.4f}")
print(f"Precision: {nb_precision:.4f}")
print(f"Recall   : {nb_recall:.4f}")
print(f"F1 Score : {nb_f1:.4f}")
print(f"AUC Score: {nb_auc:.4f}")
print(f"MCC Score: {nb_mcc:.4f}")

Gaussian Naive Bayes Performance Metrics
---------------------------------------------
Accuracy : 0.8289
Precision: 0.3440
Recall   : 0.4610
F1 Score : 0.3940
AUC Score: 0.8107
MCC Score: 0.3012


## 2.6 Random Forest Classifier

Random Forest is an ensemble Machine Learning algorithm that combines multiple Decision Trees to improve prediction accuracy and reduce overfitting. Instead of relying on a single Decision Tree, Random Forest builds several trees using different subsets of the training data and combines their predictions through majority voting.

In this assignment, the Random Forest classifier is used to predict whether a customer will subscribe to a term deposit based on customer information and previous marketing campaign details. Random Forest generally provides better generalization performance than a single Decision Tree and is widely used for classification problems due to its robustness and high predictive accuracy.

In [ ]:
# =============================================================================
# 2.6.1 Train the Random Forest Classifier
# =============================================================================

from sklearn.ensemble import RandomForestClassifier

random_forest_model = RandomForestClassifier(
    n_estimators=10,
    max_depth=8,
    random_state=42
)

# Train the model using the training dataset
random_forest_model.fit(X_train, y_train)

print("Random Forest model trained successfully.")

Random Forest model trained successfully.


In [ ]:
# =============================================================================
# 2.6.2 Predict and Evaluate the Random Forest Classifier
# =============================================================================

# Predict the target values for the testing dataset
rf_y_pred = random_forest_model.predict(X_test)

# Predict probabilities for AUC calculation
rf_y_prob = random_forest_model.predict_proba(X_test)[:, 1]

# Calculate evaluation metrics
rf_accuracy = accuracy_score(y_test, rf_y_pred)
rf_precision = precision_score(y_test, rf_y_pred)
rf_recall = recall_score(y_test, rf_y_pred)
rf_f1 = f1_score(y_test, rf_y_pred)
rf_auc = roc_auc_score(y_test, rf_y_prob)
rf_mcc = matthews_corrcoef(y_test, rf_y_pred)

# Display the evaluation metrics
print("Random Forest Performance Metrics")
print("-" * 45)
print(f"Accuracy : {rf_accuracy:.4f}")
print(f"Precision: {rf_precision:.4f}")
print(f"Recall   : {rf_recall:.4f}")
print(f"F1 Score : {rf_f1:.4f}")
print(f"AUC Score: {rf_auc:.4f}")
print(f"MCC Score: {rf_mcc:.4f}")

Random Forest Performance Metrics
---------------------------------------------
Accuracy : 0.8983
Precision: 0.6741
Recall   : 0.3034
F1 Score : 0.4185
AUC Score: 0.9064
MCC Score: 0.4072


In [ ]:
# =============================================================================
# Step 3.2: Save the Trained Machine Learning Models
# =============================================================================

import os
import joblib
import sklearn
import sys

# Create the model folder
os.makedirs("model", exist_ok=True)

# Save all trained models
joblib.dump(logistic_model, "model/logistic_regression.pkl")
joblib.dump(decision_tree_model, "model/decision_tree.pkl")
joblib.dump(knn_model, "model/knn.pkl")
joblib.dump(naive_bayes_model, "model/naive_bayes.pkl")
joblib.dump(random_forest_model, "model/random_forest.pkl")

print("All trained models have been saved successfully.")

# Display model file sizes
print("\nSaved Model Sizes:")
for file in os.listdir("model"):
    size_mb = os.path.getsize(os.path.join("model", file)) / (1024 * 1024)
    print(f"{file:<30} {size_mb:.2f} MB")

# Display environment information
print("\nEnvironment Information")
print("------------------------")
print("Python Version :", sys.version)
print("Scikit-learn Version :", sklearn.__version__)
print("Joblib Version :", joblib.__version__)

All trained models have been saved successfully.

Saved Model Sizes:
knn.pkl                        4.69 MB
decision_tree.pkl              0.45 MB
logistic_regression.pkl        0.00 MB
naive_bayes.pkl                0.00 MB
random_forest.pkl              0.28 MB

Environment Information
------------------------
Python Version : 3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0]
Scikit-learn Version : 1.6.1
Joblib Version : 1.5.3


In [ ]:
# =============================================================================
# Step 3.3 Create Sample Test Dataset
# =============================================================================

# Select the first 10 customer records (excluding the target column)
test_data = df.drop(columns=['y']).head(10)

# Save the sample data as a CSV file
test_data.to_csv("test_data.csv", index=False)

print("Sample test_data.csv created successfully.")

Sample test_data.csv created successfully.


In [ ]:
# Create a ZIP file of the model folder

import shutil

shutil.make_archive("model", "zip", "model")

print("model.zip created successfully!")

model.zip created successfully!


In [ ]:
# =============================================================================
# Step 3.3 Verify Saved Models
# =============================================================================

import joblib

models = {
    "Logistic Regression": "model/logistic_regression.pkl",
    "Decision Tree": "model/decision_tree.pkl",
    "KNN": "model/knn.pkl",
    "Naive Bayes": "model/naive_bayes.pkl",
    "Random Forest": "model/random_forest.pkl"
}

print("Verifying saved models...\n")

for name, path in models.items():
    model = joblib.load(path)
    print(f"✓ {name} loaded successfully ({type(model).__name__})")

Verifying saved models...

✓ Logistic Regression loaded successfully (LogisticRegression)
✓ Decision Tree loaded successfully (DecisionTreeClassifier)
✓ KNN loaded successfully (KNeighborsClassifier)
✓ Naive Bayes loaded successfully (GaussianNB)
✓ Random Forest loaded successfully (RandomForestClassifier)


In [ ]:
import os

print("Current Working Directory:")
print(os.getcwd())

print("\nFiles in the current directory:")
print(os.listdir())

Current Working Directory:
/content

Files in the current directory:
['.config', 'bank-full.csv', 'test_data.csv', 'model.zip', 'model', 'sample_data']


In [ ]:
# =============================================================================
# Step 3.4 Create a Test Dataset for Streamlit
# =============================================================================

# Take 20 records from the test set
test_data = X_test.copy().head(20)

# Save as CSV (without the target column)
test_data.to_csv("test_data.csv", index=False)

print("test_data.csv created successfully.")
print("Shape:", test_data.shape)

test_data.head()

test_data.csv created successfully.
Shape: (20, 16)


,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome
3776,40,1,1,1,0,580,1,0,2,16,8,192,1,-1,0,3
9928,47,7,2,1,0,3644,0,0,2,9,6,83,2,-1,0,3
33409,25,8,2,2,0,538,1,0,0,20,0,226,1,-1,0,3
31885,42,4,1,2,0,1773,0,0,0,9,0,311,1,336,1,0
15738,56,4,1,2,0,217,0,1,0,21,5,121,2,-1,0,3


In [ ]:
# =============================================================================
# Step 3.5 Test Prediction Using Saved Models
# =============================================================================

import joblib
import pandas as pd

# Load one model
model = joblib.load("model/random_forest.pkl")

# Load sample test data
test_df = pd.read_csv("test_data.csv")

# Predict
predictions = model.predict(test_df)

# Display results
results = test_df.copy()
results["Prediction"] = predictions

print("Predictions generated successfully.")
results.head()

Predictions generated successfully.


,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,Prediction
0,40,1,1,1,0,580,1,0,2,16,8,192,1,-1,0,3,0
1,47,7,2,1,0,3644,0,0,2,9,6,83,2,-1,0,3,0
2,25,8,2,2,0,538,1,0,0,20,0,226,1,-1,0,3,0
3,42,4,1,2,0,1773,0,0,0,9,0,311,1,336,1,0,0
4,56,4,1,2,0,217,0,1,0,21,5,121,2,-1,0,3,0


In [ ]:
# Test the Random Forest model on the first 20 test records

predictions = random_forest_model.predict(X_test.head(20))

print(predictions)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]


In [ ]:
# =============================================================================
# Step 4 - Model Evaluation
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    roc_auc_score,
    precision_score,
    recall_score,
    f1_score,
    matthews_corrcoef
)

models = {
    "Logistic Regression": logistic_model,
    "Decision Tree": decision_tree_model,
    "KNN": knn_model,
    "Naive Bayes": naive_bayes_model,
    "Random Forest": random_forest_model
}

evaluation_results = []

for model_name, model in models.items():

    y_pred = model.predict(X_test)

    # AUC requires probabilities if available
    if hasattr(model, "predict_proba"):
        y_prob = model.predict_proba(X_test)[:, 1]
        auc = roc_auc_score(y_test, y_prob)
    else:
        auc = roc_auc_score(y_test, y_pred)

    evaluation_results.append({
        "Model": model_name,
        "Accuracy": accuracy_score(y_test, y_pred),
        "AUC Score": auc,
        "Precision": precision_score(y_test, y_pred),
        "Recall": recall_score(y_test, y_pred),
        "F1 Score": f1_score(y_test, y_pred),
        "MCC Score": matthews_corrcoef(y_test, y_pred)
    })

evaluation_df = pd.DataFrame(evaluation_results)

evaluation_df

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but KNeighborsClassifier was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but KNeighborsClassifier was fitted without feature names
  warnings.warn(


,Model,Accuracy,AUC Score,Precision,Recall,F1 Score,MCC Score
0,Logistic Regression,0.122305,0.512564,0.120846,1.000000,0.215634,0.015098
1,Decision Tree,0.874046,0.704972,0.478182,0.482126,0.480146,0.408494
2,KNN,0.843857,0.502464,0.109489,0.041247,0.059920,-0.007474
3,Naive Bayes,0.828928,0.810730,0.344049,0.461045,0.394046,0.301205
4,Random Forest,0.898264,0.906442,0.674134,0.303391,0.418458,0.407172


In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrices = {}

models = {
    "Logistic Regression": logistic_model,
    "Decision Tree": decision_tree_model,
    "K-Nearest Neighbour": knn_model,
    "Gaussian Naive Bayes": naive_bayes_model,
    "Random Forest": random_forest_model
}

for model_name, model in models.items():

    y_pred = model.predict(X_test)

    cm = confusion_matrix(y_test, y_pred)

    confusion_matrices[model_name] = cm

    print(f"\n{model_name}")
    print(cm)

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but KNeighborsClassifier was fitted without feature names
  warnings.warn(



Logistic Regression
[[  15 7937]
 [   0 1091]]

Decision Tree
[[7378  574]
 [ 565  526]]

K-Nearest Neighbour
[[7586  366]
 [1046   45]]

Gaussian Naive Bayes
[[6993  959]
 [ 588  503]]

Random Forest
[[7792  160]
 [ 760  331]]
